### ISSUE TITLE
PrepareProviderConfig's PreparedConfig not used?

### ISSUE THREAD

**Issue by entropitor** [hashicorp/terraform#26792]
### Terraform Version

```
Terraform v0.12.29

Your version of Terraform is out of date! The latest version
is 0.13.5. You can update by downloading from https://www.terraform.io/downloads.html
```

### Expected Behavior

I'm writing a terraform SDK for NodeJS and I was expecting that the prepared config of https://github.com/hashicorp/terraform/blob/0c5430f88d04c0dfeaf4fb51c1293dd0ff3acbe1/providers/provider.go#L17-L19 would be returned to the provider in the ConfigureRequest https://github.com/hashicorp/terraform/blob/0c5430f88d04c0dfeaf4fb51c1293dd0ff3acbe1/providers/provider.go#L161-L169

### Actual Behavior
However, it seems that always the raw config is returned. Which seems to correspond with the fact the only place we seem to execute this PrepareProviderConfig request, we don't do anything with the output:
 
https://github.com/hashicorp/terraform/blob/0c5430f88d04c0dfeaf4fb51c1293dd0ff3acbe1/terraform/eval_validate.go#L101-L108

Am I missing something? Is the protocol not meant to be interpreted that way?

**Comment by jbardin** [comment]
Thanks @entropitor 

You are correct, it seems this is not currently being used! The legacy SDK takes takes care of the necessary logic within the `Configure` call as well, so this has slipped through the cracks since we have very few providers working directly against the grpc protocol.

**Comment by entropitor** [comment]
Ah okay 👍 

It's especially annoying because it seems terraform is starting multiple processes of the plugin so we really have to rerun the PrepareProviderConfig in each Configure, as caching it in the provider plugin doesn't work

**Comment by jbardin** [comment]
Yes, each use of a provider will start a new process, so they must always be independently configured. 
The reason for allowing `PrepareProviderConfig` to modify the configuration was as a concession to maintain some existing behaviors during the 0.12 transition, but that apparently the issues were resolved in another manner. While we would like to fix this and stick to the documented behavior of `PrepareProviderConfig`, you should be able to accomplish whatever is required within the `Configure` call itself, and only use `PrepareProviderConfig` for its validation purpose.

**Comment by entropitor** [comment]
Okay 👍

I guess a note on the interface could help a lot towards this. It could potentially "solve" this bug. 

Btw it's only because of this bug that I found the documentation of the different methods. The grpc proto file doesn't include the same documentation as the general go plugin interface. It could be a nice addition to copy some of the comments over to the prototype files to help explain the interface 😄  (or at least add a link to the other file on github with the documentatio)

**Comment by jbardin** [comment]
> I guess a note on the interface could help a lot towards this. It could potentially "solve" this bug.

Yeah, I'm going to do a little more research here, but if we cannot find any remaining functionality issues, it might be more prudent to simply document `PrepareProviderConfig` as being only used for validation to ensure broader compatibility.

Thanks!

**Comment by jbardin** [comment]
I went back and reviewed the legacy plugin behavior, and confirmed that the issue which `PrepareProviderConfig` was anticipating was avoided in another way. 

I've chosen for now not to change the behavior at all, since there is presently no use case for it. What I did add in #26816, was a second call to `PrepareProviderConfig` immediately before `Configure` so that the provider can validate the final config if it wasn't complete during the initial static validation. A note was also added to the interface docs that the field is not going to be used.

We are currently writing up more protocol documentation in preparation for a new SDK, and should be able to add more details for implementors in that documentation.

Thanks!
